### PR TITLE
Eliminate capture avoiding substitution

### PR DESCRIPTION
**Pull request by ajreynol**
We have adopted the policy that shadowing should be handled in the rewriter, not during substitution.

This eliminates capture avoiding substitution method, which confuses this policy.  It extends the rewriter for all closure operators to eliminate shadowing explicitly.  This should make the implementation of substitution+rewriting more robust, as we no longer have to care about the source of substitutions.

Note that the quantifiers rewriter (for FORALL, EXISTS) already eliminates shadowing during one of its passes.

### CODE REVIEW

File: src/theory/uf/theory_uf_rewriter.cpp [hunk 1]
```diff
@@ -125,28 +93,21 @@ RewriteResponse TheoryUfRewriter::postRewrite(TNode node)
...
     }
   }
   else if (k == kind::LAMBDA)
   {
     Node ret = rewriteLambda(node);
-    return RewriteResponse(REWRITE_DONE, ret);
+    if (ret != node)
+    {
+      return RewriteResponse(REWRITE_AGAIN_FULL, ret);
+    }
```
Comment (HanielB) [inline src/theory/uf/theory_uf_rewriter.cpp:110]: Why is this necessary? What new rewrites do we expect to apply once shadowing is eliminated?

File: src/theory/builtin/theory_builtin_rewriter.cpp [hunk 2]
```diff
@@ -120,6 +121,12 @@ Node TheoryBuiltinRewriter::rewriteWitness(TNode node)
     // (witness ((x Bool)) (not x)) ---> false
     return NodeManager::currentNM()->mkConst(false);
   }
+  // eliminate shadowing
+  Node retElimShadow = ElimShadowNodeConverter::eliminateShadow(node);
+  if (retElimShadow != node)
+  {
+    return retElimShadow;
+  }
   return node;
```
Comment (HanielB) [inline src/theory/builtin/theory_builtin_rewriter.cpp:130]:
```suggestion
  return ElimShadowNodeConverter::eliminateShadow(node);
```

File: src/expr/elim_shadow_converter.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+#include "expr/node_converter.h"
+
+namespace cvc5::internal {
+
+/**
+ * This converts a node into one that does not involve shadowing with the
+ * given variables. In particular, if the given vars passed to the constructor
+ * are bound in any quantifier in a subterm of the node to convert, they
+ * are replaced by fresh variables.
+ *
```
Comment (HanielB) [inline src/expr/elim_shadow_converter.h:30]: Clarify that the given variables are the variables bound by the closure given to the constructor. Also, "quantifier" -> "binder"

File: src/expr/elim_shadow_converter.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+   * bound by closure n that occurs as a subterm of closure q.
+   */
+  static Node getElimShadowVar(const Node& q, const Node& n, const Node& v);
+
+  /** Eliminate shadowing in the closure q */
+  static Node eliminateShadow(const Node& q);
+
+ private:
+  /** The quantified formula to eliminate shadowing from */
+  Node d_quant;
```
Comment (HanielB) [inline src/expr/elim_shadow_converter.h:58]: It's more general than quantifiers. You should update all references to "quant.*" in this class.

**Review by HanielB — COMMENTED** [review]
Do I understand correctly that one side effect of doing things like this is that we will *always* rename the variables of binders?

File: src/theory/uf/theory_uf_rewriter.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/uf/theory_uf_rewriter.cpp:110]: None, I guess. In general `REWRITE_AGAIN_FULL` is a way to be safe, else this is a small optimization.

File: src/expr/elim_shadow_converter.h (hunk 4 above)
Comment (ajreynol) [reply src/expr/elim_shadow_converter.h:58]: Done.

**Comment by ajreynol** [comment]
To answer your question, no, we only rename variables when they are shadowed.  In fact, for:
`(forall ((y Int)) .... (forall ((x Int) (y Int) (z Int)) ...))`
*only* the inner `y` is renamed.

**Review by HanielB — APPROVED** [review]
